README.md
`ssh_certificate_parser` is a small library for interacting with [OpenSSH host/user certificates](https://cvsweb.openbsd.org/cgi-bin/cvsweb/~checkout~/src/usr.bin/ssh/PROTOCOL.certkeys?rev=1.15&content-type=text/plain). Specifically, it supports RSA, DSA, and Ed25519 keys signed by an RSA certificate authority. It does not currently validate the CA signature, but merely parses out some fields.

![CI](https://github.com/EasyPost/ssh_certificate_parser/workflows/CI/badge.svg)
[![Documentation Status](https://readthedocs.org/projects/ssh-certificate-parser/badge/?version=latest)](https://ssh-certificate-parser.readthedocs.io/en/latest/?badge=latest)

This work is licensed under the ISC license, a copy of which can be found at [LICENSE.txt](LICENSE.txt)

## Usage

This module contains a single class, `SSHCertificate`. You can construct it with the `.from_bytes` or `.from_file` classmethods.

```python
from ssh_certificate_parser import SSHCertificate

cert = SSHCertificate.from_file('/etc/ssh/ssh_host_rsa_key-cert.pub')

remaining_seconds_of_validity = cert.remaining_validity
```

Full documentation is at <https://ssh-certificate-parser.readthedocs.io/en/latest/>.

## Development

```sh
# Install dependencies
make install

# Lint project
make lint

# Test project
make test
make coverage
```


CHANGES.md
# CHANGELOG

## v1.6.0

- Adds support for certs that last forever
- Adds support for `ed25519` certs

## v1.5.0

- Adds proper support for `ecdsa-sha2-nistp256`

## v1.4.0

- Adds proper support for `ecdsa-sha2-nistp384`

## v1.3.3

- Fix RtD link issue

## v1.3.2

- Fix PyPI project URLs

## v1.3.1

- Fix `setup.py` issues

## v1.3.0

- Can now parse ECDSA keys (if they're signed with an RSA CA)
- Add `.from_file` constructor on `SSHCertificate`
- Add a bunch of type hints
- Improve documentation a bit

## v1.2.0

- Can now parse DSA and Ed25519 keys (although they still have to have been signed by an RSA CA)
- Unit tests
- `key_type` is now in the `SSHCertificate` object
- `pubkey_parts` is a dictionary containing the appropriate parts for that key (e.g., `n` and `e` for RSA)
- Now raises subclasses of `ssh_certificate_parser.errors.SSHCertificateParserError` instead of just `ValueError` with a string description

## v1.1.0

- CA certificate fingerprint (equivalent to `ssh-keygen -l -f /path/to/key.pub`) is now parsed for ssh-rsa CAs. I don't have any ECDSA/Ed25519 CAs so I haven't tested them!

## v1.0.0

- Initial release


tests/test_basic.py
import datetime
from pathlib import Path

from ssh_certificate_parser import SSHCertificate, RSAPublicKey

import pytest


def test_rsa_cert(mocker):
    with open('tests/data/web1_rsa_key-cert.pub', 'rb') as f:
        cert = SSHCertificate.from_bytes(f.read())
    assert cert.key_type == 'ssh-rsa-cert-v01@openssh.com'
    assert cert.valid_after == datetime.datetime(2018, 10, 9, 0, 2, 41)
    assert cert.valid_before == datetime.datetime(2019, 10, 15, 0, 2, 41)
    assert cert.key_id == 'web1'
    assert cert.principals == ['web1.example.com']
    assert cert.serial == 1
    mocker.patch('ssh_certificate_parser.utcnow', return_value=datetime.datetime(2018, 10, 9, 0, 2, 39))
    assert cert.remaining_validity == 0
    mocker.patch('ssh_certificate_parser.utcnow', return_value=datetime.datetime(2018, 10, 9, 0, 2, 41))
    assert cert.remaining_validity == 32054400.0
    mocker.patch('ssh_certificate_parser.utcnow', return_value=datetime.datetime(2019, 10, 16, 0, 2, 41))
    assert cert.remaining_validity == 0
    d = cert.asdict()
    assert d['ca_fingerprint'] == 'SHA256:dHmZ5LUyoGNqkH8zzg7hndOJp9rqKRgq21OVMW7JhPk'
    assert d['cert_type'] == 'SSH2_CERT_TYPE_HOST'
    assert d['crits'] == []
    assert d['exts'] == []

    assert isinstance(cert.ca, RSAPublicKey)


def test_forever_cert(mocker):
    with open('tests/data/web4_rsa_key-cert.pub', 'rb') as f:
        cert = SSHCertificate.from_bytes(f.read())
    assert cert.valid_after == datetime.datetime(1970, 1, 1, 0, 0, 0)
    assert cert.valid_before == datetime.datetime(9999, 12, 31, 0, 0, 0)
    assert cert.key_id == 'web4'
    assert cert.principals == ['web4.example.com']
    assert cert.serial == 0
    mocker.patch('ssh_certificate_parser.utcnow', return_value=datetime.datetime(1970, 1, 2, 0, 0, 0))
    assert cert.remaining_validity > 0
    mocker.patch('ssh_certificate_parser.utcnow', return_value=datetime.datetime(2018, 2, 5, 0, 0, 0))
    assert cert.remaining_validity > 0
    mocker.patch('ssh_certificate_parser.utcnow', return_value=datetime.datetime(2099, 3, 8, 0, 0, 0))
    assert cert.remaining_validity > 0
    d = cert.asdict()
    assert d['ca_fingerprint'] == 'SHA256:sZ8QWMpND0GZa8pm3MFNV8nHB2+ssdukl/FyZ49JBgU'
    assert d['cert_type'] == 'SSH2_CERT_TYPE_USER'
    assert d['crits'] == []
    assert d['exts'] == []


def test_ed25519_cert():
    with open('tests/data/web1_ed25519_key-cert.pub', 'rb') as f:
        cert = SSHCertificate.from_bytes(f.read())
    assert cert.key_type == 'ssh-ed25519-cert-v01@openssh.com'
    assert cert.valid_after == datetime.datetime(2018, 10, 9, 0, 8, 1)
    assert cert.valid_before == datetime.datetime(2019, 10, 15, 0, 8, 1)
    assert cert.key_id == 'web1'
    assert cert.principals == ['web1.example.com']
    assert cert.serial == 1
    d = cert.asdict()
    assert d['ca_fingerprint'] == 'SHA256:dHmZ5LUyoGNqkH8zzg7hndOJp9rqKRgq21OVMW7JhPk'
    assert d['cert_type'] == 'SSH2_CERT_TYPE_HOST'
    assert d['crits'] == []
    assert d['exts'] == []


def test_dsa_cert():
    with open('tests/data/web2_dsa_key-cert.pub', 'rb') as f:
        cert = SSHCertificate.from_bytes(f.read())
    assert cert.key_type == 'ssh-dss-cert-v01@openssh.com'
    assert cert.valid_after == datetime.datetime(2018, 10, 9, 0, 11, 55)
    assert cert.valid_before == datetime.datetime(2019, 10, 15, 0, 11, 55)
    assert cert.key_id == 'web2'
    assert cert.principals == ['web2.example.com']
    assert cert.serial == 2
    d = cert.asdict()
    assert d['ca_fingerprint'] == 'SHA256:dHmZ5LUyoGNqkH8zzg7hndOJp9rqKRgq21OVMW7JhPk'
    assert d['cert_type'] == 'SSH2_CERT_TYPE_HOST'
    assert d['crits'] == []
    assert d['exts'] == []


@pytest.mark.parametrize('key_size', ['256', '384', '521'])
def test_ecdsa_cert(key_size):
    with open('tests/data/web3_{0}_ecdsa_key-cert.pub'.format(key_size), 'rb') as f:
        cert = SSHCertificate.from_bytes(f.read())
    assert cert.key_type == 'ecdsa-sha2-nistp{0}-cert-v01@openssh.com'.format(key_size)
    assert cert.valid_after == datetime.datetime(2021, 8, 12, 9, 2, 3)
    assert cert.valid_before == datetime.datetime(2022, 8, 14, 7, 59, 59)
    assert cert.key_id == 'CN=web3,CA=new_ca'
    assert cert.principals == ['web3', 'web3.example']
    assert cert.serial == 0
    d = cert.asdict()
    assert d['ca_fingerprint'] == 'SHA256:DEGzMMIkpgHKD7EdQr8p9BemrKdzhmTbhFZch4Scx1w'
    assert d['cert_type'] == 'SSH2_CERT_TYPE_HOST'
    assert d['crits'] == []
    assert d['exts'] == []
    assert d['pubkey_parts']['curve'] == 'nistp{0}'.format(key_size)


def test_from_file():
    path = 'tests/data/web1_ed25519_key-cert.pub'
    cert = SSHCertificate.from_file(path)
    assert cert.key_type == 'ssh-ed25519-cert-v01@openssh.com'

    with open(path, 'rb') as f:
        cert = SSHCertificate.from_file(f)
        assert cert.key_type == 'ssh-ed25519-cert-v01@openssh.com'

    cert = SSHCertificate.from_file(Path(path))
    assert cert.key_type == 'ssh-ed25519-cert-v01@openssh.com'


tests/test_parser_functions.py
from ssh_certificate_parser import parser_helpers
from ssh_certificate_parser.errors import InputTooShortError

import pytest


def test_take_u32():
    assert (0, b'') == parser_helpers.take_u32(b'\x00\x00\x00\x00')
    assert (1, b'\x00\x00\x00\x00') == parser_helpers.take_u32(b'\x00\x00\x00\x01\x00\x00\x00\x00')
    with pytest.raises(InputTooShortError):
        parser_helpers.take_u32(b'\x00')


def test_take_u64():
    assert (0, b'') == parser_helpers.take_u64(b'\x00\x00\x00\x00\x00\x00\x00\x00')
    assert (1, b'\x00') == parser_helpers.take_u64(b'\x00\x00\x00\x00\x00\x00\x00\x01\x00')
    with pytest.raises(InputTooShortError):
        parser_helpers.take_u64(b'\x00')


def test_take_pascal_bytestring():
    assert (b'foo', b'') == parser_helpers.take_pascal_bytestring(b'\x00\x00\x00\x03foo')
    assert (b'foo', b'\x01') == parser_helpers.take_pascal_bytestring(b'\x00\x00\x00\x03foo\x01')
    assert (b'fo\x00', b'\x01') == parser_helpers.take_pascal_bytestring(b'\x00\x00\x00\x03fo\x00\x01')
    with pytest.raises(InputTooShortError):
        parser_helpers.take_pascal_bytestring(b'\x00\x00\x00\x03fo')


def test_take_pascal_string():
    assert ('foo', b'') == parser_helpers.take_pascal_string(b'\x00\x00\x00\x03foo')
    assert ('fπø', b'\x00') == parser_helpers.take_pascal_string(b'\x00\x00\x00\x05f\xcf\x80\xc3\xb8\x00')


def test_take_list():
    assert ([1, 2], b'\x01') == parser_helpers.take_list(
        b'\x00\x00\x00\x08\x00\x00\x00\x01\x00\x00\x00\x02\x01',
        parser_helpers.take_u32
    )


ssh_certificate_parser/parser_helpers.py
import struct

from .errors import InputTooShortError


def take_u32(byte_array):
    if len(byte_array) < 4:
        raise InputTooShortError()
    return struct.unpack('!L', byte_array[:4])[0], byte_array[4:]


def take_u64(byte_array):
    if len(byte_array) < 8:
        raise InputTooShortError()
    return struct.unpack('!Q', byte_array[:8])[0], byte_array[8:]


def take_pascal_bytestring(byte_array):
    string_len, rest = take_u32(byte_array)
    if len(rest) < string_len:
        raise InputTooShortError()
    return rest[:string_len], rest[string_len:]


def take_pascal_string(byte_array):
    string_len, rest = take_u32(byte_array)
    return rest[:string_len].decode('utf-8'), rest[string_len:]


def take_list(byte_array, per_item_callback):
    overall, rest = take_pascal_bytestring(byte_array)
    lst = []
    while overall:
        item, overall = per_item_callback(overall)
        lst.append(item)
    return lst, rest


ssh_certificate_parser/__init__.py
import base64
import datetime
import enum
import hashlib
from pathlib import Path
from typing import List

import attr

from .errors import UnsupportedCertificateTypeError, UnsupportedKeyTypeError
from .parser_helpers import (take_list, take_pascal_bytestring,
                             take_pascal_string, take_u32, take_u64)

__author__ = 'EasyPost <oss@easypost.com>'
version_info = (1, 4, 0)
__version__ = '.'.join(str(s) for s in version_info)


class CertType(enum.Enum):
    SSH2_CERT_TYPE_USER = 1
    SSH2_CERT_TYPE_HOST = 2


@attr.s
class PublicKey(object):
    """The public key of a CA"""
    raw = attr.ib()

    @property
    def fingerprint(self):
        dgt = hashlib.sha256(self.raw).digest()
        b64 = base64.standard_b64encode(dgt).decode('ascii')
        return 'SHA256:{0}'.format(b64.rstrip('='))


@attr.s
class RSAPublicKey(PublicKey):
    modulus = attr.ib()
    exponent = attr.ib()


def take_rsa_cert(raw_pubkey, byte_array):
    modulus_len, byte_array = take_u32(byte_array)
    modulus = byte_array[:modulus_len]
    byte_array = byte_array[modulus_len:]
    exponent_len, byte_array = take_u32(byte_array)
    exponent = byte_array[:exponent_len]
    return RSAPublicKey(modulus=modulus, exponent=exponent, raw=raw_pubkey)


def take_ed25519_cert(raw_pubkey):
    return PublicKey(raw=raw_pubkey)


def utcnow():
    return datetime.datetime.utcnow()  # pragma: no cover


@attr.s
class SSHCertificate(object):
    """Representation of a signed SSH certificate"""
    serial: int = attr.ib()
    cert_type: CertType = attr.ib()
    key_id: str = attr.ib()
    principals: List[str] = attr.ib()
    valid_after: datetime.date = attr.ib()
    valid_before: datetime.date = attr.ib()
    crits: List[str] = attr.ib()
    exts: List[str] = attr.ib()
    ca: PublicKey = attr.ib()
    signature: bytes = attr.ib()
    key_type: str = attr.ib()
    pubkey_parts: dict = attr.ib()

    def asdict(self):
        """Return a dictionary with the important properties of this certificate"""
        dct = attr.asdict(self)
        del dct['ca']
        dct['valid_after'] = dct['valid_after'].isoformat()
        dct['valid_before'] = dct['valid_before'].isoformat()
        dct['cert_type'] = dct['cert_type'].name
        dct['signature'] = base64.b64encode(dct['signature']).decode('ascii')
        dct['ca_fingerprint'] = self.ca.fingerprint
        dct['pubkey_parts'] = dict(
            (
                k,
                base64.b64encode(v).decode('ascii')
                if isinstance(v, bytes)
                else v
            )
            for k, v in
            dct['pubkey_parts'].items()
        )
        return dct

    @classmethod
    def from_file(cls, path_or_file_object):
        """Construct a new SSHCertificate from a path or open file descriptor

        :path_or_file_object: A path or open file object pointing to a valid certifcate
        """
        if isinstance(path_or_file_object, (str, Path)):
            with open(path_or_file_object, 'rb') as f:
                return cls.from_file(f)
        return cls.from_bytes(path_or_file_object.read())

    @classmethod
    def from_bytes(cls, byte_array):
        """Construct a new SSHCertificate from a byte array

        :param byte_array: bytestring or equivalent object
        """
        if b' ' in byte_array:
            blob = byte_array.split(b' ')[1]
        else:
            blob = byte_array
        blob = base64.b64decode(blob)
        key_type, blob = take_pascal_string(blob)
        pubkey_parts = {}
        if key_type == 'ssh-rsa-cert-v01@openssh.com':
            pubkey_parts['nonce'], blob = take_pascal_bytestring(blob)
            pubkey_parts['n'], blob = take_pascal_bytestring(blob)
            pubkey_parts['e'], blob = take_pascal_bytestring(blob)
        elif key_type == 'ssh-ed25519-cert-v01@openssh.com':
            pubkey_parts['nonce'], blob = take_pascal_bytestring(blob)
            pubkey_parts['pubkey'], blob = take_pascal_bytestring(blob)
        elif key_type == 'ssh-dss-cert-v01@openssh.com':
            pubkey_parts['nonce'], blob = take_pascal_bytestring(blob)
            pubkey_parts['p'], blob = take_pascal_bytestring(blob)
            pubkey_parts['q'], blob = take_pascal_bytestring(blob)
            pubkey_parts['g'], blob = take_pascal_bytestring(blob)
            pubkey_parts['pubkey'], blob = take_pascal_bytestring(blob)
        elif key_type in (
            'ecdsa-sha2-nistp256-cert-v01@openssh.com',
            'ecdsa-sha2-nistp384-cert-v01@openssh.com',
            'ecdsa-sha2-nistp521-cert-v01@openssh.com'
        ):
            pubkey_parts['nonce'], blob = take_pascal_bytestring(blob)
            pubkey_parts['curve'], blob = take_pascal_string(blob)
            pubkey_parts['point'], blob = take_pascal_bytestring(blob)
        else:
            raise UnsupportedKeyTypeError(key_type)
        serial, blob = take_u64(blob)
        cert_type, blob = take_u32(blob)
        cert_type = CertType(cert_type)
        key_id, blob = take_pascal_string(blob)
        principals, blob = take_list(blob, take_pascal_string)
        valid_after, blob = take_u64(blob)
        valid_after = datetime.datetime.utcfromtimestamp(valid_after)
        valid_before, blob = take_u64(blob)
        # Forever certificate may be large, change to a value that can be handled by CPython
        max_python_datetime = datetime.datetime(year=9999, month=12, day=31)
        if valid_before > max_python_datetime.timestamp():
            valid_before = max_python_datetime
        else:
            valid_before = datetime.datetime.utcfromtimestamp(valid_before)
        crits, blob = take_list(blob, take_pascal_string)
        exts, blob = take_list(blob, take_pascal_string)
        unknown, blob = take_pascal_bytestring(blob)
        raw_ca, blob = take_pascal_bytestring(blob)
        ca_cert_type, raw_ca_rest = take_pascal_string(raw_ca)
        if ca_cert_type in ('ssh-rsa', 'ecdsa-sha2-nistp256', 'ecdsa-sha2-nistp384'):
            ca_cert = take_rsa_cert(raw_ca, raw_ca_rest)
        elif ca_cert_type == 'ssh-ed25519':
            ca_cert = take_ed25519_cert(raw_ca)
        else:
            raise UnsupportedCertificateTypeError(ca_cert_type)
        signature = blob
        return SSHCertificate(
            serial, cert_type, key_id, principals, valid_after, valid_before,
            crits, exts, ca_cert, signature, key_type, pubkey_parts
        )

    @property
    def remaining_validity(self):
        """Number of seconds of remaining validity on this certificate"""
        now = utcnow()
        if now > self.valid_before or now < self.valid_after:
            return 0
        else:
            return (self.valid_before - now).total_seconds()


__all__ = ['SSHCertificate', 'CertType']
